src/main/java/net/imglib2/algorithm/bspline/BSplineDecomposition.java

```diff
@@ -0,0 +1,582 @@
+
+		// recurse fwd
+		for( int i = 1; i < N; i++ )
+		{
+				
+			srcAccess.fwd( dimension );
+			destAccess.fwd( dimension );
+			
+			// c[i] = v[i] + z * c[i-1]
+			S coef = destAccess.get();
+			coef.setReal( srcAccess.get().getRealDouble() );
+			previous.mul( z );
+			coef.add( previous );
+
+			
+			previous.set( coef );
+		}
+		// here, destAccess at position N-1
+
+		previous.setReal( padOperation( srcAccess, dimension, previous.getRealDouble(), z, Ci ));
+
+		/*
+		 * After calling this method:
+		 *   destAccess at position N-2
+		 *   previous holds the value of coef at N-1
+		 */
+		for( long i = N-1; i >= 0; i-- )
+		{
+			// coefs[ i ] = Z1 * ( coefs[i+1] - coefs[ i ]);
+			// 		      = -Z1 * ( coefs[i] - coefs[ i + 1 ]);
+			S coef = destAccess.get();
+			coef.sub( previous );
+			coef.mul( -z );
+			previous.set( coef );
+
+			srcAccess.bck( dimension );
+			destAccess.bck( dimension );
+		}
+
+		// TODO may need to back things up to make sure we're in the right place
+
+	}
+
+	private < R extends RealType<R>> double padOperation( RandomAccess< R > access, int dimension, double last, double z, double Ci )
+	{
+		access.fwd( dimension );	
+		padding[ 0 ] = access.get().getRealDouble() + z * last;
+
+		// do paddingWidth more causal operations
+		for( int i = 1; i < paddingWidth; i++ )
+		{
+			access.fwd( dimension );	
+			padding[ i ] = access.get().getRealDouble() + z * padding[ i - 1 ];
+		}
+
+		// initialize anticausal
+		// coefs[ N-1 ] = Ci * ( coefs[ N-1 ] + z * coefs[ N-2 ] );
+		padding [ paddingWidth - 1 ] += z * padding [ paddingWidth - 2 ];
+		padding [ paddingWidth - 1 ] *= Ci;
+
+		// do paddingWith anticausal operations
+		for( int i = paddingWidth - 2; i >=0; i-- )
+		{
+			padding[ i ] = z * ( padding[ i + 1 ] - padding[ i ] );
+		}
+
+		return padding[ 0 ];
+	}
+
+	/*
+	 * The positions of the data and coefs RandomAccess must be set correctly
+	 * before calling this.
+	 * 
+	 * This method has side effects:
+	 *   calls coefs.bck( 0 )
+	 *   changes the value of previous
+	 */
+	protected static <T extends RealType<T>> void initializeAntiCausalCoefficients(
+			final double z,
+			final double c,
+			final T previous, // temporary variable
+			final RandomAccess<T> coefs )
+	{
+		T last = coefs.get(); // last has the value at coefs[ N-1 ]
+
+		coefs.bck( 0 );
+		previous.set( coefs.get() ); // previous has the value at coefs[ N-2 ]
+
+		// coefs[ N-1 ] = Ci * ( coefs[ N-1 ] + z * coefs[ N-2 ] );
+		previous.mul( z );
+		last.add( previous );
+		last.mul( c );
+
+		previous.set( last );
+	}
+	
+	/**
+	 * Data must be 1d or permuted such that the first dimension
+	 * is the dimension to be iterated over.
+	 * 
+	 * See Box 2 of Unser 1999
+	 */
+	protected static <T extends RealType<T>> double initializeCausalCoefficients(
+			final double z,
+			final double tolerance,
+			final int dimension,
+			final RandomAccess<T> dataAccess )
+	{
+		// TODO this method is wasteful if the interval is zero-extended
```

Comment:
Still a TODO?

Reply:
its not a TODO, but rather a comment.  So TODO removed.